### PR TITLE
RawTx API 

### PR DESCRIPTION
**Pull request by nagarajm22**
## Summary

- Implementing Raw Tx API for ocean

## RPCs

- RawTx  `send` Api   

## Implications

- Storage
  - [ ] Database reindex required
  - [ ] Database reindex optional
  - [ ] Database reindex not required
  - [ ] None

- Consensus
  - [ ] Network upgrade required
  - [ ] Includes backward compatible changes
  - [ ] Includes consensus workarounds
  - [ ] Includes consensus refactors
  - [ ] None

### CODE REVIEW

File: lib/ain-ocean/src/api/rawtx.rs [hunk 1]
```diff
@@ -1,27 +1,152 @@
...
-    format!("Details of raw transaction with txid {}", txid)
+async fn validate(hex: String) {
+    if !hex.starts_with("040000000001") {
+        return;
+    }
+    let buffer = hex::decode(hex).expect("Decoding failed");
+    let transaction: Transaction = deserialize(&buffer).expect("Failed to deserialize transaction");
+    if transaction.output.len() != 2 {
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (Jouzo) [reply lib/ain-ocean/src/api/rawtx.rs:142]: Does not need to be async either

[the remark this answers is not in the document]
Comment (Jouzo) [reply lib/ain-ocean/src/api/rawtx.rs:142]: Guess you took this from jellyfish :
```

  async validate (hex: string): Promise<void> {
    if (!hex.startsWith('040000000001')) {
      return
    }

    const buffer = SmartBuffer.fromBuffer(Buffer.from(hex, 'hex'))
    const transaction = new CTransactionSegWit(buffer)

    if (transaction.vout.length !== 2) {
      return
    }

    if (transaction.vout[0].script.stack.length !== 2) {
      return
    }

    if (transaction.vout[0].script.stack[0].type !== OP_CODES.OP_RETURN.type) {
      return
    }

    if ((transaction.vout[0].script.stack[1] as OP_DEFI_TX).tx.type !== CCompositeSwap.OP_CODE) {
      return
    }

    const dftx = (transaction.vout[0].script.stack[1] as OP_DEFI_TX).tx.data as CompositeSwap
    if (dftx.pools.length === 0) {
      return
    }

    const lastPoolId = dftx.pools[dftx.pools.length - 1].id
    const toTokenId = `${dftx.poolSwap.toTokenId}`

    const info = await this.deFiDCache.getPoolPairInfo(`${lastPoolId}`)
    if (info?.idTokenA === toTokenId || info?.idTokenB === toTokenId) {
      return
    }
    throw new BadRequestApiException('Invalid CompositeSwap')
  }
  ```
  
  Why are we skipping the rest of the validation ?

[the remark this answers is not in the document]
Comment (nagarajm22) [reply lib/ain-ocean/src/api/rawtx.rs:142]: I took the reference form it, as we are not using `CTransactionSegWit ` just using `Transaction`

File: lib/ain-ocean/src/api/rawtx.rs [hunk 2]
```diff
@@ -1,27 +1,148 @@
...
+    Error, Result,
+};
+
+#[ocean_endpoint]
+async fn send_rawtx(
+    Extension(ctx): Extension<Arc<AppContext>>,
+    Json(raw_tx_dto): Json<RawTxDto>,
+) -> Result<String> {
+    println!("{:?}", raw_tx_dto.hex.clone());
+    validate(raw_tx_dto.hex.clone())?;
```
Comment (canonbrother) [inline lib/ain-ocean/src/api/rawtx.rs:29]:
```suggestion
    validate(&raw_tx_dto.hex)?;
```

File: lib/ain-ocean/src/api/rawtx.rs [hunk 3]
```diff
@@ -1,27 +1,148 @@
...
+    let mut max_fee = Some(default_max_fee_rate().unwrap().to_sat());
+    if let Some(fee_rate) = raw_tx_dto.max_fee_rate {
+        let sat_per_bitcoin = Decimal::new(100_000_000, 0);
```
Comment (canonbrother) [inline lib/ain-ocean/src/api/rawtx.rs:32]:
```suggestion
        let sat_per_bitcoin = dec!(100_000_000);
```

File: lib/ain-ocean/src/api/rawtx.rs [hunk 4]
```diff
@@ -1,27 +1,148 @@
...
+        let fee_in_satoshis = fee_rate * sat_per_bitcoin;
```
Comment (canonbrother) [inline lib/ain-ocean/src/api/rawtx.rs:33]:
```suggestion
        let fee_in_satoshis = fee_rate.checked_mul(sat_per_bitcoin);
```

File: lib/ain-ocean/src/api/rawtx.rs [hunk 5]
```diff
@@ -1,27 +1,148 @@
...
+        max_fee = fee_in_satoshis.round().to_u64();
+    }
+    let trx = defichain_rpc::RawTx::raw_hex(raw_tx_dto.hex);
+    match ctx.client.send_raw_transaction(trx, max_fee).await {
```
Comment (canonbrother) [inline lib/ain-ocean/src/api/rawtx.rs:37]: hmm.. is this needed? `defichain_rpc::RawTx::raw_hex()`??
```suggestion
    match ctx.client.send_raw_transaction(raw_tx_dto.hex, max_fee).await {
```

File: lib/ain-ocean/src/api/rawtx.rs [hunk 6]
```diff
@@ -1,27 +1,148 @@
...
+        fee: tx_result.fee.map(|amount| amount.to_sat()),
+        details,
+        hex: tx_result.hex,
+    };
+    Ok(Response::new(raw_tx))
 }
 
-async fn get_rawtx(Path(txid): Path<String>) -> String {
-    format!("Details of raw transaction with txid {}", txid)
+fn validate(hex: String) -> Result<()> {
```
Comment (canonbrother) [inline lib/ain-ocean/src/api/rawtx.rs:131]: missing dftx and composite swap tx checks
```
if ((transaction.vout[0].script.stack[1] as OP_DEFI_TX).tx.type !== CCompositeSwap.OP_CODE) {
      return
    }

    const dftx = (transaction.vout[0].script.stack[1] as OP_DEFI_TX).tx.data as CompositeSwap
    if (dftx.pools.length === 0) {
      return
    }

    const lastPoolId = dftx.pools[dftx.pools.length - 1].id
    const toTokenId = `${dftx.poolSwap.toTokenId}`

    const info = await this.deFiDCache.getPoolPairInfo(`${lastPoolId}`)
    if (info?.idTokenA === toTokenId || info?.idTokenB === toTokenId) {
      return
    }
    throw new BadRequestApiException('Invalid CompositeSwap')
```

File: lib/ain-ocean/src/api/rawtx.rs [hunk 7]
```diff
@@ -1,27 +1,148 @@
...
 }
 
 pub fn router(ctx: Arc<AppContext>) -> Router {
+    println!("{:?}", ctx.network);
     Router::new()
         .route("/send", post(send_rawtx))
-        .route("/test", get(test_rawtx))
-        .route("/:txid", get(get_rawtx))
+        .route("/test", post(test_rawtx))
+        .route("/:txid", get(get_raw_tx))
```
Comment (canonbrother) [inline lib/ain-ocean/src/api/rawtx.rs:146]: hmm.. can standardize the naming??
```suggestion
        .route("/test", post(test_rawtx))
        .route("/:txid", get(get_raw_tx))
```

File: lib/ain-ocean/src/lib.rs [hunk 8]
```diff
@@ -34,6 +34,7 @@ pub mod storage;
 use crate::storage::ocean_store::OceanStore;
 
 pub type Result<T> = std::result::Result<T, Error>;
+pub type ApiResult<T> = std::result::Result<T, ApiError>;
```
Comment (canonbrother) [inline lib/ain-ocean/src/lib.rs:37]: hmm... why do we need this?? is this unused??

File: lib/ain-ocean/src/api/rawtx.rs [hunk 9]
```diff
@@ -1,27 +1,148 @@
...
+};
+
+#[ocean_endpoint]
+async fn send_rawtx(
+    Extension(ctx): Extension<Arc<AppContext>>,
+    Json(raw_tx_dto): Json<RawTxDto>,
+) -> Result<String> {
+    println!("{:?}", raw_tx_dto.hex.clone());
+    validate(raw_tx_dto.hex.clone())?;
+    let mut max_fee = Some(default_max_fee_rate().unwrap().to_sat());
```
Comment (canonbrother) [inline lib/ain-ocean/src/api/rawtx.rs:30]: should check `raw_tx_dto.max_fee` is some/none first then else default max fee rate

File: lib/ain-ocean/src/model/raw_tx.rs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+#[serde(rename_all = "camelCase")]
+pub struct RawTxDto {
+    pub hex: String,
+    pub max_fee_rate: Option<Decimal>,
+}
+
+pub fn default_max_fee_rate() -> Option<Amount> {
+    let default_max_fee_rate = Amount::from_btc(0.1).unwrap_or_default();
+    Some(default_max_fee_rate)
+}
```
Comment (canonbrother) [inline lib/ain-ocean/src/model/raw_tx.rs:15]: i think the `Decimal` is good enough to fulfill the bitcoin::Amount
just make it simple

**Comment by canonbrother** [comment]
test suite failed test checked:
`should throw BadRequestError due to invalid txn`
```
test › should throw BadRequestError due to invalid txn
    expect(received).toStrictEqual(expected) // deep equality
    - Expected  - 3
    + Received  + 4
      Object {
        "at": Any<Number>,
    -   "code": 400,
    -   "message": "Transaction decode failed",
    -   "type": "BadRequest",
    +   "code": 404,
    +   "message": "TX decode failed",
    +   "type": "NotFound",
    +   "url": "/v0/regtest/rawtx/test",
 ```  
 can catch the errmsg `Transaction decode failed` then throw `BadRequest` error
 same apply with `send › should throw BadRequestException due to invalid txn`

File: lib/ain-ocean/src/api/rawtx.rs [hunk 11]
```diff
@@ -1,27 +1,148 @@
...
+            eprintln!("Failed to send raw transaction: {:?}", e);
+            Err(Error::RpcError(e))
+        }
+    }
 }
 
-async fn test_rawtx() -> String {
-    "Testing raw transaction".to_string()
+#[ocean_endpoint]
+async fn get_raw_tx(
```
Comment (canonbrother) [inline lib/ain-ocean/src/api/rawtx.rs:79]: test failed

File: lib/ain-ocean/src/api/rawtx.rs (hunk 7 above)
Comment (nagarajm22) [reply lib/ain-ocean/src/api/rawtx.rs:146]: Its same as in `jf`

File: lib/ain-ocean/src/lib.rs (hunk 8 above)
Comment (nagarajm22) [reply lib/ain-ocean/src/lib.rs:37]: this need to remove

File: lib/ain-ocean/src/model/raw_tx.rs (hunk 10 above)
Comment (nagarajm22) [reply lib/ain-ocean/src/model/raw_tx.rs:15]: I reference from https://github.com/DeFiCh/rust-defichain-rpc/blob/main/integration_test/src/main.rs 
`static ref FEE: Amount = Amount::from_btc(0.001).unwrap();`

File: lib/ain-ocean/src/api/rawtx.rs (hunk 11 above)
Comment (nagarajm22) [reply lib/ain-ocean/src/api/rawtx.rs:79]: This is bcoz of `jf` `model RawTrasnction` is different compare what is return from `rust_defi_rpc`

File: lib/ain-ocean/src/api/rawtx.rs (hunk 7 above)
Comment (canonbrother) [reply lib/ain-ocean/src/api/rawtx.rs:146]: i mean the function naming either we go for `test_rawtx` or `test_raw_tx`.. we need to standardize the naming

File: lib/ain-ocean/src/model/raw_tx.rs (hunk 10 above)
Comment (canonbrother) [reply lib/ain-ocean/src/model/raw_tx.rs:15]: can just do `const MAX_FEE_RATE = dec!(0.1)` as simple as possible

File: lib/ain-ocean/src/api/rawtx.rs (hunk 6 above)
Comment (nagarajm22) [reply lib/ain-ocean/src/api/rawtx.rs:144]: Issue is deserialising the `hex` string to `defichain_rpc::defichain_rpc_json::{blockchain::Transaction}`. Is there any implementation ?

Comment (canonbrother) [reply lib/ain-ocean/src/api/rawtx.rs:144]: try refer
https://github.com/DeFiCh/ain/blob/e3cb077a2e51210e93618ba6d92f5b873e523901/lib/ain-ocean/src/indexer/mod.rs#L146-L177

File: lib/ain-ocean/src/api/rawtx.rs (hunk 11 above)
Comment (canonbrother) [reply lib/ain-ocean/src/api/rawtx.rs:92]: https://github.com/BirthdayResearch/jellyfishsdk/blob/cb1c46722c9495dcfffd1e9f8c6d9a684d4dc7c9/apps/whale-api/src/module.api/rawtx.controller.ts#L99

**Comment by canonbrother** [comment]
test is still failed

```
FAIL apps/whale-api/src/module.api/__defid__/rawtx.controller.defid.ts (13.697 s)
  test
    ✓ should accept valid txn (23 ms)
    ✓ should accept valid txn with given maxFeeRate (20 ms)
    ✕ should throw BadRequestError due to invalid txn (6 ms)
    ✓ should throw BadRequestError due to high fees (17 ms)
  send
    ✕ should send valid txn and validate tx out (17 ms)
    ✕ should send valid txn with given maxFeeRate and validate tx out (17 ms)
    ✕ should throw BadRequestException due to invalid txn (2 ms)
    ✕ should throw BadRequestException due to high fees (18 ms)
  get
    ✕ should accept valid txn and return hex (26 ms)
    ✕ should throw NotFoundException due to tx id not found (3 ms)
```

File: lib/ain-ocean/src/api/rawtx.rs (hunk 7 above)
Comment (canonbrother) [reply lib/ain-ocean/src/api/rawtx.rs:210]: its not fixed

File: lib/ain-ocean/src/error.rs [hunk 12]
```diff
@@ -70,6 +70,10 @@ pub enum Error {
     TryFromIntError(#[from] std::num::TryFromIntError),
     #[error(transparent)]
     Other(#[from] anyhow::Error),
+    #[error("Validation error: {0}")]
+    ValidationError(String),
+    #[error("Validation error: {0}")]
```
Comment (Jouzo) [inline lib/ain-ocean/src/error.rs:75]: Wrong description

Comment (nagarajm22) [reply lib/ain-ocean/src/error.rs:75]: already update it :)

**Review by canonbrother — APPROVED** [review]

**Comment by canonbrother** [comment]
hmm.. test failed due to `Connection refused`.. re-run CI and check again...

**Review by canonbrother — APPROVED** [review]

**Review by Jouzo — APPROVED** [review]

**Pull request merged**